### ISSUE TITLE
"Comments-> Set..." does not consistently default to the same type of comment

### ISSUE THREAD

**Issue by wothke** [NationalSecurityAgency/ghidra#3921]
**Describe the bug**
When right-clicking in the "decompiled code view" and using the "Comments-> Set..." function, I expect to add a comment at the line that I am at (i.e. a comment that actually shows up in that view afterward).

I guess that this means that the respective dialog should default to the "Pre comment" mode. (I've also encountered situations where a respective "Pre comment" just doesn't show in the decompiled code view - but I'll leave that for another bug report)

However the respective comment-editing dialog seems to randomly select the mode that it defaults to; sometimes it is "Pre comment" sometimes it it "Plate comment". This is just very annoying.

I have the impression that the problem can most easily be reproduced when the 1st line of code within a decompiled function is involved. 

I notice a secondary issue that may well be related: I've added a "Pre comment" to the first line of code within a decompiled function and that comment correctly shows above that line in the decompiled code view. When I then right-click that comment to edit it via "Comments -> Set..."
the respective dialog incorrectly default to "Plate comment" and I first have to manually switch to
"Pre comment" to get to the comment text that I set out to modify. This makes no sense.. the context menu should know what type of comment I clicked on and automatically select the correct one for editing.

**Environment (please complete the following information):**
 - OS: Win10
 - Ghidra Version: 10.1.1

**Comment by Pokechu22** [comment]
This is actually documented behavior:

https://github.com/NationalSecurityAgency/ghidra/blob/7b38e60298b3ffabdf628887ba4ed2fc3e0bd936/Ghidra/Features/Decompiler/src/main/doc/decompileplugin.xml#L4107-L4112

Experimentally, it seems like local variable declarations and the empty line between them and the code _also_ count as part of the function header (which sorta makes sense, as they don't directly map to any instructions in the function).  A plate comment for them goes above the function, while a pre comment goes inside of the function, but after all of the local variable declarations.    I don't know if there's a good way of changing this.

The secondary issue you mention is the same logic applying for comments: a comment in the function's header (even if it's a pre comment that shows up right before the code) is considered part of the function's header, and a plate comment in the function's body is considered part of the function's body.  Regardless of the comment's type, the only factor is if the comment is considered to be part of the header or not; if their display is enabled, the same thing happens with EOL comments, post comments, and repeatable comments (which show up in the disassembly wherever something is referenced - I didn't know that until just now).  But changing it to edit the currently selected comment doesn't sound conceptually hard, at least.

**Comment by wothke** [comment]
> But changing it to edit the currently selected comment doesn't sound conceptually hard, at least.

The logic for selecting the anchor of a new comment is one thing.. but when the user specifcially clicks on an existing comment 
to edit it, then it seems obvious which comment should be selected for editing by default.

**Comment by Pokechu22** [comment]
I did a bit of further investigation, and it seems like the non-decompiler version of "Comments -> Set..." already behaves this way.  Here's its code:

https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Features/Base/src/main/java/ghidra/app/plugin/core/comments/CommentsActionFactory.java#L121-L137

On the other hand, here's the decompiler's version:

https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Features/Decompiler/src/main/java/ghidra/app/plugin/core/comments/DecompilerCommentsActionFactory.java#L119-L144

The last parameter of [`CommentType.getCommentType`](https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/util/CommentType.java#L31-L60) is the default type to use, so perhaps the decompiler's plate/pre logic should instead set that parameter instead of directly changing the return value (but I'm not 100% sure that would actually work).

**Comment by Pokechu22** [comment]
The challenge here is that the non-decompiler version knows when the cursor is over a [`CommentFieldLocation`](https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/util/CommentFieldLocation.java), but the decompiler version always gets a [`DecompilerLocation`](https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Features/Decompiler/src/main/java/ghidra/app/decompiler/DecompilerLocation.java) which has a [`ClangToken`](https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Features/Decompiler/src/main/java/ghidra/app/decompiler/ClangToken.java) that could be a [`ClangCommentToken`](https://github.com/NationalSecurityAgency/ghidra/blob/3eb03b1608b24aaae1cd1d5c86a5df36270ac54d/Ghidra/Features/Decompiler/src/main/java/ghidra/app/decompiler/ClangCommentToken.java), but that doesn't know what kind of comment it originally was.

My earlier suggested change does partially work (if you have a comment selected, and it's the only comment at an address, then that comment will be the one that gets edited), but breaks if there are multiple comments at the same address (with a preferred order instead of choosing the comment you've got selected).  To fix this issue the decompiler would also need to store the type of comment.

There's also some weirdness with multiple addresses mapping to the same line, but that's actually handled correctly if you select the comment - the `ClangCommentToken` knows which address it corresponds to.  You can get semi-weird results if you select the line after the comment, as depending on the selected token you'll edit different lines, but this is accurately reflected by the cursor's position in the disassembly.

**Comment by Pokechu22** [comment]
#3974 will fix this, though the implementation is fairly ugly.

**Comment by strensk** [comment]
Just trying to understand if what I'm seeing is bug or feature. Plate vs Pre vs Post vs EOL work as expected in disassembler, but all appear as Pre in decompiler. Sounds like maybe this is acknowledged as current situation, and somewhere down in the work queue. Correct?

**Comment by Pokechu22** [comment]
No, that's a separate thing.  This issue is about clicking on a comment in the decompiler and then trying to edit it not using the same type of comment as the selected comment (and to a lesser extent the comment type chosen when editing it varying for other reasons).  The comments all appearing the same way in the decompiler is different, and IMO something that would be difficult to change (as multiple instructions all get combined into one statement in the decompiler, so post/pre/EOL are a lot harder to reason with, though plate could be handled separately).
